types/bootstrap-menu/index.d.ts

```diff
@@ -0,0 +1,114 @@
+
+    /**
+     * Obtain specific data about the currently opened element, to pass it to the rest of user-defined functions of an action.
+     */
+    fetchElementData?: (jqueryElement: JQuery) => BootstrapMenuRowElement;
+
+    /**
+     * Array or object containing the list of actions to be rendered in the context menu.
+     */
+    actions: { [key: string]: BootstrapMenuActions } | ReadonlyArray<BootstrapMenuActions>;
+
+    /**
+     * Array to group actions to render them next to each other, with a separator between each group.
+     */
+    actionsGroups?: ReadonlyArray<ReadonlyArray<string>>;
+
+    /**
+     *  message to show when there are no actions to show in a menu
+     * (isShown() returned false on all actions)
+     */
+    noActionsMessage?: string;
+
+    /**
+     * In some weird cases, another plugin may be installing 'click' listeners
+     * in the anchors used for each action of the context menu, and stopping
+     * the event bubbling before it reachs this plugin's listener.
+     *
+     * For those cases, _actionSelectEvent can be used to change the event we
+     * listen to, for example to 'mousedown'.
+     *
+     * Unless the context menu is not working due to this and a workaround is
+     * needed, this option can be safely ignored.
+     */
+    _actionSelectEvent?: string;
+}
+
+/**
+ * Every function attribute is called before rendering the menu each time it is opened.
+ * If `fetchElementData` was defined, these functions will receive as first argument its returned value for the currently selected element.
+ */
+export interface BootstrapMenuActions {
+    /**
+     * The name of the action.
+     */
+    name: string;
+
+    /**
+     * Handler to run when an action is clicked.
+     */
+    onClick: (rowElement: BootstrapMenuRowElement) => void;
+
+    /**
+     * Optional, Font Awesome class of the icon to show for the action.
+     */
+    iconClass?: string;
+
+    /**
+     * Optional, classes to add to the action.
+     */
+    classNames?: string | object | (() => string | object);
+
+    /**
+     * Optional, decides if the action should be shown or hidden in the context menu.
+     */
+    isShown?: (rowElement: BootstrapMenuRowElement) => boolean;
+
+    /**
+     * Optional, decides if the action should appear enabled or disabled in the context menu.
+     */
+    isEnabled?: (rowElement: BootstrapMenuRowElement) => boolean;
+}
+
+export type BootstrapMenuSource = "mouse" | "element";
+
+export type BootstrapMenuPosition = "aboveLeft" | "aboveRight" | "belowLeft" | "belowRight";
+
+export type BootstrapMenuEvent = "click" | "right-click" | "hover";
+
+export type BootstrapMenuRowElement = any;
+
+/**
+ * The function to mount the menu to an existing UI item
+ * @param selector Css selector of the element to attach the Menu to
+ * @param options Options to configure the menu
+ */
+export default function BootstrapMenu(selector: string, options: BootstrapMenuOptions): void;
```

Comment:
I believe this is incorrect as the package doesn't export a `default`. Instead you should use `export  =` syntax. `export = BootstrapMenu`.

Also is not a `function` but a class.

```ts
declare class BootstrapMenu {
    constructor(...)
}
```

Last, the rest of the exports can me added to a namespace:
```ts
declare namespace BoostrapMenu {
    export interface BootstrapMenuOptions {
        ...
    }
}
```

You can see more details on the [template documentation](https://www.typescriptlang.org/docs/handbook/declaration-files/templates/module-class-d-ts.html).

Reply:
https://github.com/dgoguerra/bootstrap-menu/blob/master/src/BootstrapMenu.js#L431 For me this looks like a default export. If I'm not doing an export default I can't do directly new BootstrapMenu with the `import BootstrapMenu = require("bootstrap-menu")`

Reply:
Ok my bad I just say that doing `export =` instead of `export class` works.